shotgun_api3/shotgun.py

```diff
@@ -1623,6 +1622,276 @@ def update_project_last_accessed(self, project, user=None):
+                                  ["sg_status_list", "is", "ip"]] } }
+          
+          The general syntax for this data structure is:
+          { entity_type: { filter_operator: all|any,
+                           filters: shotgun_filter_list } }
+            
+          The filter_operator key is optional and if not specified, 'all' 
+          will be used.
+        
+        A dictionary with keys terms and matches will be returned:
+        
+        {'matches': [{'id': 734,
+                      'type': 'Asset',
+                      'name': 'Bunny',
+                      'project_id': 65,                      
+                      'image': 'https://...',
+                      'links': ['', ''],
+                      'status': 'fin'},
+                      
+                      {'id': 558,
+                       'type': 'Task'
+                       'name': 'FX',
+                       'project_id': 65,
+                       'image': 'https://...',
+                       'links': ['Shot', 'bunny_010_0010'],
+                       'status': 'fin'}],
+            'terms': ['bunny']}
+        
+        The links field will contain information about any linked entity. 
+        This is useful when for example presenting tasks and you want to 
+        display what shot or asset the task is associated with.
+
+        :param text: Text to search for. This must be at least three 
+                     characters long, or an exception will be raised.
+        :param entity_types: Dictionary to specify which entity types to search 
+                             across. See above for usage examples.
+        :param project_ids: List of projects to search. By default, all 
+                            projects will be searched. 
+        :param limit: Specify the maximum number of matches to return.
+        
+        :returns: A complex dictonary structure, see above for example.
+        """        
+        if self.server_caps.version and self.server_caps.version < (6, 2, 0):
+                raise ShotgunError("auto_complete requires server version 6.2.0 or "\
+                    "higher, server is %s" % (self.server_caps.version,))
+        
+        # convert entity_types structure into the form 
+        # that the API endpoint expects
+        if not isinstance(entity_types, dict):
+            raise ValueError("entity_types parameter must be a dictionary")
+        
+        api_entity_types = {}
+        for (entity_type, filter_dict) in entity_types.iteritems():
+            
+            if not isinstance(filter_dict, dict):
+                raise ValueError("value of entity_types['%s'] must "
+                                 "be a dictionary" % entity_type)
+            
+            if "filters" in filter_dict:
+                # there is a filter query inside the dict. 
+                # convert it into native form 
+                filter_operator = filter_dict.get("filter_operator")
```

Comment:
Is there a reason to make this specific to the dictionary type filter syntax?  Seems like it would be easier to explain and document if we just accept (per entity type key) the same structures we support for general filter arguments... ie lists in the form [["name", "contains", "foo"]] or the dictionary syntax.  Then we can just update the docs to refer to the general page on how to filter.